### ISSUE TITLE
Extension File Filtering

### ISSUE THREAD

**Issue by glasserre-pdw** [FileZ/FileZ#113]
Hi

First, thanks for this very useful application.

I try to add a file extension filtering on uploaded file. (Example: _.mp3;_.avi)

I read the source code and i think that the filter must be place on Upload.php in the startAction function, but i don't know how implement the filtering.

```
  public function startAction () {
        $this->secure ();
        fz_log ('uploading');
        fz_log ('uploading', FZ_LOG_DEBUG, $_FILES);
        $response = array (); // returned data

        // check if request exceed php.ini post_max_size
        if ($_SERVER ['CONTENT_LENGTH'] > $this->shorthandSizeToBytes (
                                                   ini_get ('post_max_size'))) {
            fz_log ('upload error (POST request > post_max_size)', FZ_LOG_ERROR);
            return $this->onFileUploadError (UPLOAD_ERR_INI_SIZE);
        }
        else if ($_FILES ['file']['error'] === UPLOAD_ERR_OK) {
            if ($this->checkQuota ($_FILES ['file'])) // Check user quota first
                return $this->onFileUploadError (UPLOAD_ERR_QUOTA_EXCEEDED);

        if ($this->checkQuota ($_FILES ['file']))
        return $this->onFileUploadError (UPLOAD_ERR_QUOTA_EXCEEDED);    

[Filtering process here]

            // Still no error ? we can move the file to its final destination
            $file = $this->saveFile ($_POST, $_FILES ['file']);
            if ($file !== null) {
                $this->sendFileUploadedMail ($file);
                return $this->onFileUploadSuccess ($file);

            } else { // Errors happened while saving or moving the uploaded file
                return $this->onFileUploadError ();
            }
        } else { // Errors happened during file upload
            return $this->onFileUploadError ($_FILES ['file']['error']);
        }
    }
```

Someone can help me ?

Best regards.

**Comment by ianbogda** [comment]
Hye,

if I should do something like this : 

1/ add a directive in [app] inside the config file

2/ check before to know if th file is in good way to be donloaded (before UserQuota…)

3/ Add a javascript check before to send the file

It could be something like below :

```
diff --git a/app/controllers/Upload.php b/app/controllers/Upload.php
index cdf6825..77a27d6 100644
--- a/app/controllers/Upload.php
+++ b/app/controllers/Upload.php
@@ -3,6 +3,7 @@
 define ('UPLOAD_ERR_QUOTA_EXCEEDED', 99);
 define('UPLOAD_ERR_VIRUS_FOUND', 100);
 define('UPLOAD_ERR_ANTIVIRUS', 101);
+define('UPLOAD_ERR_ALLOWED_EXTS', 102);

 /**
  * Controller used to upload files and monitor progression
@@ -19,6 +20,9 @@ class App_Controller_Upload extends Fz_Controller {
         fz_log ('uploading', FZ_LOG_DEBUG, $_FILES);
         $response = array (); // returned data

+       // First of all check if file's type is allowed
+       if ($this->extensionNotAllowed()) return $this->onFileUploadError (UPLOAD_ERR_ALLOWED_EXTS);
+
         // check if request exceed php.ini post_max_size
         if ($_SERVER ['CONTENT_LENGTH'] > $this->shorthandSizeToBytes (
                                                    ini_get ('post_max_size'))) {
@@ -54,6 +58,9 @@ class App_Controller_Upload extends Fz_Controller {
         $response = array (); // returned data
         option('visitor',true);

+       // First of all check if file's type is allowed
+       if ($this->extensionNotAllowed()) return $this->onFileUploadError (UPLOAD_ERR_ALLOWED_EXTS);
+
         // check if request exceed php.ini post_max_size
         if ($_SERVER ['CONTENT_LENGTH'] > $this->shorthandSizeToBytes (
                                                    ini_get ('post_max_size'))) {
@@ -296,6 +303,27 @@ class App_Controller_Upload extends Fz_Controller {
         return $this->returnData ($response);
     }

+   /**
+    * Function called to check if file's type is allowed or not to be downloaded
+    *
+    */ 
+   private function extensionNotAllowed() {
+        $allowed_exts = ( fz_config_get ('app', 'allowed_extensions') ) ? fz_config_get ('app', 'allowed_extensions') : '';
+
+       // No extension restriction
+       if ('' === $allowed_exts) return false;
+
+       // Extension restriction
+       // Check extension
+       $allowed_exts = explode(',', $allowed_exts);
+       $extension = end(explode('.', $_FILE['file']['name']));
+
+       // TODO : add chack with mime-type
+       if (in_array($extension, $allowed_exts)) return false;
+
+       return true;
+   }
+
     /**
      * Function called on file upload error. A message corresponding to the error
      * code passed as parameter is return to the user. Error codes come from
@@ -337,6 +365,9 @@ class App_Controller_Upload extends Fz_Controller {
             case UPLOAD_ERR_QUOTA_EXCEEDED:
                 $response ['statusText'] .= __r('You exceeded your disk space quota (%space%).',
                     array ('space' => fz_config_get ('app', 'user_quota')));
+           case UPLOAD_ERR_ALLOWED_EXTS:
+               $response ['statusText'] .= __r('The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%.',
+                   array ('allowed_exts' => fz_config_get ('app', 'allowed_exts')));
         }
         return $this->returnData ($response);
     }
diff --git a/app/views/main/_upload_form.php b/app/views/main/_upload_form.php
index 4c8ad79..cbcfabf 100644
--- a/app/views/main/_upload_form.php
+++ b/app/views/main/_upload_form.php
@@ -1,3 +1,20 @@
+  <?php // Check file extensions
+       $matches = (fz_config_get ('app', 'allowed_exts') ) ? fz_config_get ('app', 'allowed_exts') : '';
+       if ('' !== $matches) {
+           $matches = explode(',', $matches);
+           $matches = implode(')|(', $matches);
+           $matches = '/\.(' . $matches . ')$/'; ?>
+
+  <script>
+  function checkExt(e) {
+    value=e.value;
+   if( !value.match(<?php echo $matches; ?>) ){
+     alert(<?php echo __r('The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%.',
+                         array ('allowed_exts' => fz_config_get ('app', 'allowed_exts'))); ?>)
+   }
+  }
+  </script>
+  <?php } ?>
   <form method="POST" enctype="multipart/form-data" action="<?php echo url_for ('upload') ?>" id="upload-form">
   <input type="hidden" name="MAX_FILE_SIZE" value="<?php echo $max_upload_size ?>" />
   <div id="file">
@@ -42,8 +59,9 @@
     </li>
   </ul>
   <div id="upload">
-    <input type="submit" id="start-upload" name="upload" class="awesome blue large" value="&raquo; <?php echo __('Upload') ?>" />
+    <input type="submit" id="start-upload" name="upload" class="awesome blue large" value="&raquo; <?php echo __('Upload') ?>" <?php if ('' !==      $matches) echo "onClick=\"checkExt(this);\""; ?>/>
     <div id="upload-loading"  style="display: none;"></div>
     <div id="upload-progress" style="display: none;"></div>
   </div>
   </form>
+
diff --git a/config/filez.ini.example b/config/filez.ini.example
index 2ec40c1..0e29324 100644
--- a/config/filez.ini.example
+++ b/config/filez.ini.example
@@ -44,6 +44,9 @@ user_factory_class = Fz_User_Factory_Ldap
 ; Max disk spaced per user. Support shorthand format : M (Mega), G (Giga)
 user_quota = 2G

+; File extension allowed to be uploaded. Empty means no restriction
+allowed_exts = mp3,avi,doc,docx
+
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ; Cron Job
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
diff --git a/i18n/de/LC_MESSAGES/default.po b/i18n/de/LC_MESSAGES/default.po
index 2a9a6fa..3c6739d 100644
--- a/i18n/de/LC_MESSAGES/default.po
+++ b/i18n/de/LC_MESSAGES/default.po
@@ -505,6 +505,10 @@ msgstr "Es wurde keine Datei hochgeladen."
 msgid "You exceeded your disk space quota (%space%)."
 msgstr "Der verfügbare Festplattenspeicher wurde überschritten (%space%)."

+#: app/controllers/Upload.php:275
+msgid "The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%."
+msgstr "Die Datei darf nicht hochgeladen werden. Beachten Sie, dass Dateien erlaubt sein müssen %allowed_exts%"
+
 #: FileZ/app/controllers/File.php:69
 msgid "Lifetime extended"
 msgstr "Dauer wurde verlängert"
diff --git a/i18n/en/LC_MESSAGES/default.po b/i18n/en/LC_MESSAGES/default.po
index a46faeb..c2e3216 100644
--- a/i18n/en/LC_MESSAGES/default.po
+++ b/i18n/en/LC_MESSAGES/default.po
@@ -474,6 +474,10 @@ msgstr ""
 msgid "You exceeded your disk space quota (%space%)."
 msgstr ""

+#: app/controllers/Upload.php:275
+msgid "The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%."
+msgstr ""
+
 #: app/controllers/File.php:69
 msgid "Lifetime extended"
 msgstr ""
diff --git a/i18n/es/LC_MESSAGES/default.po b/i18n/es/LC_MESSAGES/default.po
index 09acf51..1b567ac 100644
--- a/i18n/es/LC_MESSAGES/default.po
+++ b/i18n/es/LC_MESSAGES/default.po
@@ -505,6 +505,10 @@ msgstr "No se subió ningún archivo"
 msgid "You exceeded your disk space quota (%space%)."
 msgstr "Has sobrepasado la cuota de espacio en disco (%space%)."

+#: app/controllers/Upload.php:275
+msgid "The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%."
+msgstr "El archivo no puede ser cargado. Tenga en cuenta que los archivos deseados necesitan estar %allowed_exts%"
+
 #: FileZ/app/controllers/File.php:69
 msgid "Lifetime extended"
 msgstr "Fecha de caducidad ampliada"
diff --git a/i18n/fr/LC_MESSAGES/default.po b/i18n/fr/LC_MESSAGES/default.po
index 8e91d73..84f2cca 100644
--- a/i18n/fr/LC_MESSAGES/default.po
+++ b/i18n/fr/LC_MESSAGES/default.po
@@ -487,6 +487,10 @@ msgstr "Aucun fichier n'a été envoyé."
 msgid "You exceeded your disk space quota (%space%)."
 msgstr "Vous avez dépassé votre quota disque (%space%)."

+#: app/controllers/Upload.php:275
+msgid "The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%."
+msgstr "Les seules extensions acceptées sont %allowed_exts%."
+
 #: FileZ/app/controllers/File.php:69
 msgid "Lifetime extended"
 msgstr "Durée de vie prolongée"
diff --git a/i18n/it/LC_MESSAGES/default.po b/i18n/it/LC_MESSAGES/default.po
index e945853..8865deb 100644
--- a/i18n/it/LC_MESSAGES/default.po
+++ b/i18n/it/LC_MESSAGES/default.po
@@ -504,6 +504,10 @@ msgstr "Nessun file è stato caricato"
 msgid "You exceeded your disk space quota (%space%)."
 msgstr "Hai superato la tua quota di spazio su disco (%space%)."

+#: app/controllers/Upload.php:275
+msgid "The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%."
+msgstr "Il file non può essere caricato. Si noti che i file devono essere autorizzati %allowed_exts%"
+
 #: FileZ/app/controllers/File.php:69
 msgid "Lifetime extended"
 msgstr "Scadenza rimandata"
diff --git a/i18n/pt/LC_MESSAGES/default.po b/i18n/pt/LC_MESSAGES/default.po
index 6b23ae3..afb0261 100644
--- a/i18n/pt/LC_MESSAGES/default.po
+++ b/i18n/pt/LC_MESSAGES/default.po
@@ -498,6 +498,10 @@ msgstr "Nenhum ficheiro enviado."
 msgid "You exceeded your disk space quota (%space%)."
 msgstr "Excedeu a sua quota de disco (%space%)."

+#: app/controllers/Upload.php:275
+msgid "The file is not allowed to be uploaded. Note that files allowed need to be %allowed_exts%."
+msgstr "O arquivo não está autorizado a ser carregado. Note que os ficheiros têm de ser autorizados %allowed_exts%"
+
 #: FileZ/app/controllers/File.php:69
 msgid "Lifetime extended"
 msgstr "Vida útil prolongada"

```

**Comment by glasserre-pdw** [comment]
Hi Ianbogda

Big thank you for your help.

I try it and tell you ;)

**Comment by glasserre-pdw** [comment]
I apply the change but :
- _upload_form.php file doesn't exist (we use UAPV-FileZ-2.0.2-17-g7524512.tar.gz package). i use the index.php instead like this :

```
<div id="upload">
    <input type="submit" id="start-upload" name="upload" class="awesome blue large" value="&raquo; <?php echo __('Upload') ?>" />
    <div id="upload-loading"  style="display: none;"></div>
    <div id="upload-progress" style="display: none;"></div>
  </div>
  </form>


+  <?php $matches = fz_config_get ('app', 'allowed_exts');
+      $matches = explode(',', $matches);
+       $matches = implode(')|(', $matches);
+      $matches = '/\.(' . $matches . ')$/'; ?>
+
+ <script>
+  function checkExt(e) {
+    value=e.value;
+   if( !value.match(<?php echo $matches; ?>) ){
+     alert("wrong extension");
+   }
+  }
+  </script>

</section>

<h2 id="uploaded-files-title"><?php echo __('Uploaded files') ?></h2>
<section id="uploaded-files">
```
- in /app/controllers/Upload.php, we just have : 

```
  define ('UPLOAD_ERR_QUOTA_EXCEEDED', 99);
+ define('UPLOAD_ERR_ALLOWED_EXTS', 102);
```

No error appears (debug=true in file.ini) and if i try to up an mp3 file, the file is added to db.

If i understand correctly the code, all the mecanism and parameter for the extension filtering are correctly added in the app but the javascript function checkExt is not call before upload.

**Comment by glasserre-pdw** [comment]
My comment was incomplete :
I complet the change with the fix 5357d54 (call of checkExt on upload).

If i up an mp3 file for example, i have the error message "Wrong Extension" but the file is up to db
Same problem with an allowed extension : Error "Wrong extension" and file is up to db.

**Comment by ianbogda** [comment]
bug fixes in https://github.com/ianbogda/FileZ/commit/49fa318715e4a88bd6c103a203927dfa62bda5ce
